src/ht.c

```diff
@@ -1,239 +1,184 @@
-		return;
-	rehash = 1;
-	ht->size_index = new_size_index;
-	ht->size = hash_sizes[ht->size_index].size;
-	ht->rehash = hash_sizes[ht->size_index].rehash;
-	ht->max_entries = hash_sizes[ht->size_index].max_entries;
-	ht->entries = 0;
-	ht->deleted_entries = 0;
-	for (e = old_ht.table; e != old_ht.table + old_ht.size; e++) {
-		if (entry_is_present (e))
-			ht_insert (ht, e->hash, e->data, e->iter);
-	}
-	free (old_ht.table);
-rehash = 0;
+SdbHash* ht_new(HashFunction hashfunction, ListComparator comparator,
+		DupKey keydup, DupValue valdup, HtKvFreeFunc pair_free) {
+	ut64 initial_size = 1024;
+	return internal_ht_new (initial_size, hashfunction, comparator, keydup,
+			valdup, pair_free);
 }
 
-SdbHash* ht_new(SdbListFree f) {
-	SdbHash *ht = R_NEW (SdbHash);
-	if (!ht) return NULL;
-	// TODO: use slices here
-	ht->list = ls_new ();
-	ht->list->free = f;
-	ht->size = hash_sizes[0].size;
-	ht->table = calloc (ht->size, sizeof (*ht->table));
-	if (!ht->table) {
-		free (ht);
-		return NULL;
+void ht_destroy(SdbHash* ht) {
+	ut32 i;
+	for (i = 0; i < ht->size; ++i) {
+		ls_free (ht->table[i]);
 	}
-	ht->size_index = 0;
-	ht->entries = 0;
-	ht->deleted_entries = 0;
-	ht->rehash = hash_sizes[ht->size_index].rehash;
-	ht->max_entries = hash_sizes[ht->size_index].max_entries;
-	return ht;
+	free (ht->table);
+	free (ht);
 }
 
-void ht_free(SdbHash *ht) {
-	if (ht) {
-		free (ht->table);
-		ls_free (ht->list);
-		free (ht);
+// Increases the size of the hashtable by 2.
+static void internal_ht_grow(SdbHash* ht) {
+	SdbHash* ht2;
+	SdbHash swap;
+	SdbKv* kvp;
+	SdbListIter* iter;
+	ut32 i;
+	ht2 = internal_ht_new (ht->size * 2, ht->hashfn, ht->cmp, ht->dupkey,
+			ht->dupvalue, (HtKvFreeFunc)ht->table[0]->free);
+	for (i = 0; i < ht->size; ++i) {
+		ls_foreach (ht->table[i], iter, kvp) {
+			(void)ht_insert (ht2, kvp->key, kvp->value);
+		}
 	}
+	// And now swap the internals.
+	swap = *ht;
+	*ht = *ht2;
+	*ht2 = swap;
+	ht_destroy (ht2);
 }
 
-void *ht_lookup(SdbHash *ht, ut32 hash) {
-	SdbHashEntry *entry;
-	if (!ht || !ht->list || !ht->list->head) return NULL;
-	entry = ht_search (ht, hash);
-	return entry? entry->data : NULL;
+// Inserts the key value pair key, value into the hashtable.
+// if update is true, allow for updates, otherwise return false if the key
+// already exists.
+static bool internal_ht_insert(SdbHash* ht, bool update, const char* key,
+		const char* value) {
+	SdbKv* kvp;
+	ut64 hash;
+	ut64 bucket;
+	bool found = true;
+	if (update) {
+		ht_delete (ht, key);
+	} else {
+		(void)ht_find (ht, key, &found);
+	}
+	if (update || !found) {
+		kvp = calloc (1, sizeof (SdbKv *));
+		if (kvp) {
+			if (ht->dupkey) {
+				kvp->key = ht->dupkey (key);
+			} else {
+				kvp->key = (char*)key;
+			}
+			if (ht->dupvalue) {
+				kvp->value = ht->dupvalue (value);
+			} else {
+				kvp->value = (char*)value;
+			}
+			hash = ht->hashfn (key);
+			bucket = hash % ht->size;
+			kvp->expire = 0;
+			kvp->value_len = strlen (kvp->value);
```

Comment:
isnt dupvalue() strlening it already?


> On 22 Jun 2016, at 16:40, Jeffrey Crowell <notifications@github.com> wrote:
> 
> In src/ht.c <https://github.com/radare/sdb/pull/102#discussion_r68065162>:
> 
> > +		kvp = calloc (1, sizeof (SdbKv *));
> > +		if (kvp) {
> > +			if (ht->dupkey) {
> > +				kvp->key = ht->dupkey (key);
> > +			} else {
> > +				kvp->key = (char*)key;
> > +			}
> > +			if (ht->dupvalue) {
> > +				kvp->value = ht->dupvalue (value);
> > +			} else {
> > +				kvp->value = (char*)value;
> > +			}
> > +			hash = ht->hashfn (key);
> > +			bucket = hash % ht->size;
> > +			kvp->expire = 0;
> > +			kvp->value_len = strlen (kvp->value);
> no it isnt???
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/radare/sdb/pull/102/files/df49a0196bf8ddbd83849180f565fa7dbe9920e9#r68065162>, or mute the thread <https://github.com/notifications/unsubscribe/AA3-lrEn_kVf1Y6FJLJ1MzqZ-CK3fmFFks5qOUldgaJpZM4I6WGf>.
>